### PR TITLE
Add Identity decorator prototype

### PR DESCRIPTION
**Pull request by markstory**
It will quack like the authentication one, but won't inherit from it because we don't want to add cross library dependencies. I tried to think of how we could support an interface that extends from cakephp/authentication and couldn't think of an approach that didn't require multiple inheritance (which we don't have).

This is just a sketch to get discussions started and try and move forwards more with how the decorator should work.

### CODE REVIEW

File: src/AuthorizationServiceInterface.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+{
+    /**
+     * Check whether the provided user can perform an action on a resource.
+     *
+     * @param \Authorization\IdentityInterface $user The user to check permissions for.
+     * @param string $action The action/operation being performed.
+     * @param mixed $resource The resource being operated on.
+     * @return bool
+     */
+    public function can(IdentityInterface $user, $action, $resource);
```
Comment (markstory) [inline src/AuthorizationServiceInterface.php]: Not sure if the service should typehint here.

Comment (robertpustulka) [reply src/AuthorizationServiceInterface.php]: I'd relax this parameter and allow anything to be passed.

File: src/IdentityDecorator.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * Delegate unknown methods to decorated identity.
+     *
+     * @param string $method The method being invoked.
+     * @param arry $args The arguments for the method.
+     * @return mixed
+     */
+    public function __call($method, $args)
+    {
+        $call = [$this->identity, $method];
```
Comment (robertpustulka) [inline src/IdentityDecorator.php]: Needs a `is_object` check probably, as the identity may be an array.

File: src/IdentityDecorator.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        $this->authorization = $service;
+        $this->identity = $identity;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function can($action, $resource)
+    {
+        return $this->authorization->can($this, $action, $resource);
```
Comment (robertpustulka) [inline src/IdentityDecorator.php]: Should `$this` be passed to the service, or maybe `$this->identity`?
Policies may typehint some internal identity object or do `instance_of` checks.

File: src/AuthorizationServiceInterface.php (hunk 1 above)
Comment (markstory) [reply src/AuthorizationServiceInterface.php]: Done!

File: src/IdentityDecorator.php (hunk 3 above)
Comment (markstory) [reply src/IdentityDecorator.php]: That is a good point. I was concerned that passing the wrapped identity would make combining permissions/policy checks defined in other resources harder. For example passing the unwrapped user would make something like the following not possible.

```php
function update($user, $thing)
{
  return $user->can('update', $this->category);
}
```

File: src/IdentityDecorator.php (hunk 2 above)
Comment (markstory) [reply src/IdentityDecorator.php]: done.

File: src/IdentityDecorator.php (hunk 3 above)
Comment (robertpustulka) [reply src/IdentityDecorator.php]: We can have some kind of `getOriginal` method in the identity. This way we could combine checks in policies and still access the original object for some internal checks.

For example:
```php
function canUpdate($identity, $object) {
   return $identity->can(‘access’, $object) || $identity->getOriginal() instanceof Admin);
}

Comment (markstory) [reply src/IdentityDecorator.php]: I can add a getOriginal() method to the interface and decorator.

**Pull request merged**